### PR TITLE
Add get port functionality

### PR DESCRIPTION
**Pull request by yabirgb**

### CODE REVIEW

**Comment by birdsarah** [comment]
Are you still working on this? If yes, change the title to have [WIP] at the front and when you're ready for review, just remove the [WIP].

**Comment by yabirgb** [comment]
@birdsarah I believe everything is ready

**Review by birdsarah — CHANGES_REQUESTED** [review]
I'd like to take a slightly different approach to the one you've built here in order to reuse code and keep the behavior of the library, and API, consistent across methods.

See the following code from get_stripped_url:

```
    purl = urlparse(url)
    _scheme = purl.scheme

    # To handle the case where we have no scheme, but we have a port
    # we have the following heuristic. Does scheme have a . in it
    # which is stdlib behavior when not recognizing a netloc due to
    # lack of //. If TLDExtract, can find a suffix in the _scheme
    # then it's probably a domain without an http.
    if '.' in _scheme:
        # From the docs: "urlparse recognizes a netloc only
        # if it is properly introduced by ‘//’". So we
        # prepend to get results we expect.
        if extractor(_scheme).suffix != '' or is_ip_address(_scheme):
            url = '//{url}'.format(url=url)
```

This could be a private method, something like, `_update_url_to_handle_port_and_no_scheme` that returns the url either modified or not.

This new private method can be used by both `get_port` and `get_stripped_url`.

This will have the nice benefit of making `get_stripped_url` a little cleaner.

Get port should then have an easy time returning the port.

I would also like to omit the `strict` flag unless you have a strong reason for it. If you do, then that reasoning should also be incorporated into the get_stripped_url function. So I'd rather open an issue to discuss the flag and then implement in both places.

I'm thrilled to see lots of tests, but we don't need to overdo it. My general rule is that I don't test imported libraries, especially not stdlib. So you only need sufficient tests to test the logic that you're implementing. If that doesn't make sense let me know.

**Comment by birdsarah** [comment]
You may also be interested in the additional comment notes I wrote here: https://github.com/mozilla/domain_utils/pull/22/files#diff-01e7661674408a06f910cdb3bd537536L200-L210

**Comment by yabirgb** [comment]
I've proceeded as you mentioned abstracting this behave in a shared function. I've found the case of urls like `localhost:5000` that were not properly catch under the old code and now they are.

My intention with those tests is to ensure that it behaves as I expect and not to test urllib. If you don't see the need for those they can be removed.

About the `strict` param my intention was to give the option to work as urllib does or under our logic. I've removed as it might be unnecessary and probably not recommended.

**Comment by birdsarah** [comment]
> My intention with those tests is to ensure that it behaves as I expect and not to test urllib. If you don't see the need for those they can be removed.

That sounds great. It's a bit of an odd library. Lots of what seem like redundant test cases but we need to make sure we've tested the edge cases we know of. Just wanted to set expectations.

**Review by birdsarah — APPROVED** [review]

File: domain_utils/domain_utils.py [hunk 1]
```diff
@@ -32,6 +32,30 @@ def is_ip_address(hostname):
         return False
 
 
+@_load_and_update_extractor
```
Comment (birdsarah) [inline domain_utils/domain_utils.py]: This decorator belongs on the main method `get_port`.

File: domain_utils/domain_utils.py [hunk 2]
```diff
@@ -264,3 +275,22 @@ def get_scheme(url, no_scheme=NO_SCHEME):
         return scheme
     else:
         return no_scheme
+
+
+def get_port(url):
```
Comment (birdsarah) [inline domain_utils/domain_utils.py]: Should be able to pass in an extractor as with other methods.

File: domain_utils/domain_utils.py [hunk 3]
```diff
@@ -219,7 +231,6 @@ def get_stripped_url(url, scheme=False, drop_non_http=False, use_netloc=True, ex
         else:
             return url
 
-    purl = urlparse(url)
```
Comment (birdsarah) [inline domain_utils/domain_utils.py:222]: Nice catch.

File: tests/test_get_port.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+def test_port_in_url():
+    assert get_port('example.com:80/path/to/index.html') == 80
+
+
+def test_port_in_domain():
+    assert get_port('example.com:5000') == 5000
+
+
+def test_port_in_domain_and_path():
```
Comment (birdsarah) [inline tests/test_get_port.py]: i think this test is redundant

File: tests/test_get_port.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+def test_port_in_ip_no_path():
+    assert get_port('10.0.0.1:80') == 80
+
+
+def test_port_no_tld():
+    assert get_port('localhost:8000') == 8000
+
+
+def test_url_with_protocol():
```
Comment (birdsarah) [inline tests/test_get_port.py:28]: could adding a standard http test as that's the most common scenario and we handle ws under a flag elsewhere (or will in #22 i should say)

**Review by birdsarah — CHANGES_REQUESTED** [review]

File: domain_utils/domain_utils.py (hunk 2 above)
Comment (birdsarah) [reply domain_utils/domain_utils.py]: and then the extractor should be passed along to the new private method.

File: domain_utils/domain_utils.py [hunk 6]
```diff
@@ -194,20 +218,8 @@ def get_stripped_url(url, scheme=False, drop_non_http=False, use_netloc=True, ex
...
-    # which is stdlib behavior when not recognizing a netloc due to
-    # lack of //. If TLDExtract, can find a suffix in the _scheme
-    # then it's probably a domain without an http.
-    if '.' in _scheme:
-        # From the docs: "urlparse recognizes a netloc only
-        # if it is properly introduced by ‘//’". So we
-        # prepend to get results we expect.
-        if extractor(_scheme).suffix != '' or is_ip_address(_scheme):
-            url = '//{url}'.format(url=url)
+    url = _adapt_url_for_port_and_scheme(url)
```
Comment (birdsarah) [inline domain_utils/domain_utils.py]: extractor should be passed along.

File: domain_utils/domain_utils.py (hunk 2 above)
Comment (birdsarah) [reply domain_utils/domain_utils.py]: you don't need to add a test for this in this PR, because I just wrote a test for the same thing in #22 which i will repurpose to test this new private method. (we'll get this PR merged before #22).

**Comment by birdsarah** [comment]
Thanks @yabirgb!
